File: vgcbench/sim.py
import json
import os
import random
import selectors
import shutil
import subprocess
import time
from collections import deque
from concurrent.futures import Future, ThreadPoolExecutor
from pathlib import Path
from typing import Protocol, TypedDict


REPO_ROOT = Path(__file__).resolve().parents[1]
TIMER_RUNNER = Path(__file__).with_name("showdown_timer.cjs")


class PlayerOpts(TypedDict):
    name: str
    team: str


class Outcome(TypedDict):
    winner: str | None
    turns: int
    log: list[str]
    pov: dict[str, list[str]]
    errors: dict[str, int]
    fallbacks: dict[str, int]


class Agent(Protocol):
    def act(self, request: dict, ctx: dict) -> str: ...

    def observe(self, pov_lines: list[str]) -> None: ...


def default_ps_dir() -> Path:
    return Path(os.environ.get("VGCBENCH_PS", REPO_ROOT / "pokemon-showdown")).expanduser()


def default_node() -> str:
    configured = os.environ.get("VGCBENCH_NODE")
    if configured:
        return str(Path(configured).expanduser())
    return shutil.which("node") or str(Path("~/.local/bin/node").expanduser())


def route_update_lines(
    lines: list[str],
    *,
    pov: dict[str, list[str]],
    log: list[str],
    pending_split: list[str],
    winner: str | None,
    turns: int,
) -> tuple[str | None, int]:
    if pending_split:
        lines = pending_split + lines
        pending_split.clear()
    index = 0
    while index < len(lines):
        line = lines[index]
        if line.startswith("|split|"):
            owner = line.split("|", 2)[2]
            if index + 2 >= len(lines):
                pending_split.extend(lines[index:])
                return winner, turns
            secret = lines[index + 1]
            public = lines[index + 2]
            if owner in pov:
                pov[owner].append(secret)
                log.append(secret)
                other = "p2" if owner == "p1" else "p1"
                if public:
                    pov[other].append(public)
            index += 3
            continue
        for pid in pov:
            pov[pid].append(line)
        log.append(line)
        if line.startswith("|turn|"):
            turns = int(line.split("|", 2)[2])
        elif line.startswith("|win|"):
            winner = line.split("|", 2)[2]
        elif line == "|tie":
            winner = None
        index += 1
    return winner, turns


class SimBattle:
    def __init__(
        self,
        format_id: str,
        p1: PlayerOpts,
        p2: PlayerOpts,
        seed=None,
        ps_dir=None,
        node=None,
    ):
        self.format_id = format_id
        self.players = {"p1": p1, "p2": p2}
        if isinstance(seed, int):
            rng = random.Random(seed)
            seed = [rng.randrange(0x10000) for _ in range(4)]
        self.seed = seed
        self.ps_dir = Path(ps_dir).expanduser() if ps_dir else default_ps_dir()
        self.node = str(Path(node).expanduser()) if node else default_node()

    def run(self, agents: dict[str, Agent]) -> Outcome:
        proc = subprocess.Popen(
            [self.node, str(TIMER_RUNNER), str(self.ps_dir)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            cwd=self.ps_dir,
        )
        if proc.stdin is None or proc.stdout is None or proc.stderr is None:
            proc.kill()
            raise RuntimeError("failed to open simulator pipes")

        recent_inputs: deque[str] = deque(maxlen=4)
        log: list[str] = []
        pov = {"p1": [], "p2": []}
        pov_cursor = {"p1": 0, "p2": 0}
        errors = {"p1": 0, "p2": 0}
        fallbacks = {"p1": 0, "p2": 0}
        last_request: dict[str, dict | None] = {"p1": None, "p2": None}
        retry_count = {"p1": 0, "p2": 0}
        pending_error: dict[str, str | None] = {"p1": None, "p2": None}
        suppress_request = {"p1": False, "p2": False}
        action_pool = ThreadPoolExecutor(max_workers=4, thread_name_prefix="vgcbench-player")
        pending_actions: dict[Future[str], str] = {}
        winner = None
        turns = 0
        last_activity = time.monotonic()

        pending_split: list[str] = []

        def send(line: str) -> None:
            nonlocal last_activity
            recent_inputs.append(line)
            proc.stdin.write((line + "\n").encode())
            proc.stdin.flush()
            last_activity = time.monotonic()

        def schedule_act(pid: str, request: dict, error: str | None = None) -> None:
            if pid in pending_actions.values():
                raise RuntimeError(f"received another request while {pid} is still deciding")
            lines = pov[pid][pov_cursor[pid] :]
            pov_cursor[pid] = len(pov[pid])
            future = action_pool.submit(agents[pid].act, request, {"pov_lines": lines, "error": error})
            pending_actions[future] = pid

        def send_finished_actions() -> None:
            for future, pid in list(pending_actions.items()):
                if not future.done():
                    continue
                del pending_actions[future]
                choice = future.result()
                send(f">{pid} {choice}")

        start = {"formatid": self.format_id}
        if self.seed is not None:
            start["seed"] = self.seed
        send(f">start {json.dumps(start, separators=(',', ':'))}")
        for pid in ("p1", "p2"):
            send(f">player {pid} {json.dumps(self.players[pid], separators=(',', ':'))}")

        selector = selectors.DefaultSelector()
        selector.register(proc.stdout, selectors.EVENT_READ)
        selector.register(proc.stderr, selectors.EVENT_READ)
        buffer = bytearray()
        stderr_buffer = bytearray()
        def route_update(lines: list[str]) -> None:
            nonlocal winner, turns
            winner, turns = route_update_lines(
                lines,
                pov=pov,
                log=log,
                pending_split=pending_split,
                winner=winner,
                turns=turns,
            )

        def handle_sideupdate(lines: list[str]) -> None:
            pid = lines[0]
            for line in lines[1:]:
                if line.startswith("|error|"):
                    errors[pid] += 1
                    retry_count[pid] += 1
                    pending_error[pid] = line
                    if retry_count[pid] >= 3:
                        send(f">{pid} default")
                        fallbacks[pid] += 1
                        suppress_request[pid] = "[Unavailable choice]" in line
                        pending_error[pid] = None
                    elif "[Unavailable choice]" not in line and last_request[pid] is not None:
                        schedule_act(pid, last_request[pid], line)
                        pending_error[pid] = None
                elif line.startswith("|timer|"):
                    for future, owner in list(pending_actions.items()):
                        if owner == pid:
                            del pending_actions[future]
                            future.cancel()
                            abandon = getattr(agents[pid], "abandon_decision", None)
                            if callable(abandon):
                                abandon()
                    if line == "|timer|autodefault":
                        fallbacks[pid] += 1
                    pov[pid].append(line)
                    log.append(line)
                elif line.startswith("|request|"):
                    payload = line[len("|request|") :]
                    if not payload:
                        continue
                    request = json.loads(payload)
                    last_request[pid] = request
                    if suppress_request[pid]:
                        suppress_request[pid] = False
                        continue
                    error = pending_error[pid]
                    if error is None:
                        retry_count[pid] = 0
                    pending_error[pid] = None
                    if not request.get("wait"):
                        schedule_act(pid, request, error)

        try:
            ended = False
            while not ended:
                send_finished_actions()
                marker = buffer.find(b"\n\n")
                if marker < 0:
                    remaining = 60 - (time.monotonic() - last_activity)
                    wait_for = 0.1 if pending_actions else max(0, remaining)
                    events = selector.select(wait_for)
                    if not events:
                        if pending_actions:
                            continue
                        raise TimeoutError(
                            f"simulator produced no output for 60 seconds after {list(recent_inputs)}"
                        )
                    chunk = b""
                    stdout_ready = False
                    for key, _ in events:
                        if key.fileobj is proc.stderr:
                            stderr_chunk = os.read(proc.stderr.fileno(), 65536)
                            if stderr_chunk:
                                stderr_buffer.extend(stderr_chunk)
                            else:
                                selector.unregister(proc.stderr)
                        else:
                            chunk = os.read(proc.stdout.fileno(), 65536)
                            stdout_ready = True
                    if not stdout_ready:
                        continue
                    if not chunk:
                        if buffer:
                            message = bytes(buffer)
                            buffer.clear()
                        else:
                            break
                    else:
                        buffer.extend(chunk)
                        last_activity = time.monotonic()
                        continue
                else:
                    message = bytes(buffer[:marker])
                    del buffer[: marker + 2]
                lines = message.decode().splitlines()
                if not lines:
                    continue
                if lines[0] == "update":
                    route_update(lines[1:])
                elif lines[0] == "sideupdate" and len(lines) >= 2:
                    handle_sideupdate(lines[1:])
                elif lines[0] == "end":
                    if len(lines) > 1:
                        data = json.loads("\n".join(lines[1:]))
                        winner = data.get("winner") or winner
                        turns = data.get("turns", turns)
                    ended = True
                elif lines[0].startswith("|"):
                    route_update(lines)
            proc.stdin.close()
            return_code = proc.wait(timeout=10)
            if not ended:
                stderr_buffer.extend(proc.stderr.read())
                stderr = stderr_buffer.decode(errors="replace").strip()[-4000:]
                raise RuntimeError(f"simulator exited {return_code} without an end block: {stderr}")
        except BaseException:
            proc.kill()
            proc.wait()
            raise
        finally:
            selector.close()
            action_pool.shutdown(wait=True, cancel_futures=True)

        for pid, agent in agents.items():
            remaining = pov[pid][pov_cursor[pid] :]
            if remaining:
                agent.observe(remaining)

        return {
            "winner": winner,
            "turns": turns,
            "log": log,
            "pov": pov,
            "errors": errors,
            "fallbacks": fallbacks,
        }

File: src/sim.ts
import type { BattleStream } from 'pokemon-showdown';
import { defaultPsDir } from './paths.js';
import type { RecoveryGate } from './recovery.js';
import { loadShowdown } from './showdown.js';
import type { TimerEvent } from './timer.js';
import { DEFAULT_TIMER_SCALE, TimerAdapter } from './timer.js';
import type { BattleAgent, BattleOutcome, BattleRequest, Pid, PlayerOptions, TimerScale } from './types.js';

interface RouteState {
  pov: Record<Pid, string[]>;
  log: string[];
  publicLog: string[];
  pendingSplit: string[];
  winner: string | null;
  turns: number;
}

export function routeUpdateLines(lines: string[], state: RouteState): void {
  if (state.pendingSplit.length) {
    lines = [...state.pendingSplit, ...lines];
    state.pendingSplit.length = 0;
  }
  for (let index = 0; index < lines.length; ) {
    const line = lines[index]!;
    if (line.startsWith('|split|')) {
      if (index + 2 >= lines.length) {
        state.pendingSplit.push(...lines.slice(index));
        return;
      }
      const owner = line.split('|', 3)[2] as Pid;
      const secret = lines[index + 1]!;
      const publicLine = lines[index + 2]!;
      if (owner === 'p1' || owner === 'p2') {
        state.pov[owner].push(secret);
        state.log.push(secret);
        if (publicLine) state.publicLog.push(publicLine);
        if (publicLine) state.pov[owner === 'p1' ? 'p2' : 'p1'].push(publicLine);
      }
      index += 3;
      continue;
    }
    state.pov.p1.push(line);
    state.pov.p2.push(line);
    state.log.push(line);
    state.publicLog.push(line);
    if (line.startsWith('|turn|')) state.turns = Number(line.slice(6));
    else if (line.startsWith('|win|')) state.winner = line.slice(5);
    else if (line === '|tie') state.winner = null;
    index += 1;
  }
}

interface PendingAction {
  token: symbol;
  promise: Promise<ActionEvent>;
}

interface ActionEvent {
  kind: 'action';
  pid: Pid;
  token: symbol;
  choice?: string;
  error?: unknown;
}

interface StreamEvent {
  kind: 'stream';
  message: string | null;
}

function streamEvent(stream: BattleStream): Promise<StreamEvent> {
  return stream.read().then((message) => ({ kind: 'stream', message: message ?? null }));
}

export class SimBattle {
  readonly seed?: [number, number, number, number];
  readonly psDir: string;

  constructor(
    readonly format: string,
    readonly players: Record<Pid, PlayerOptions>,
    seed?: number | [number, number, number, number],
    psDir = defaultPsDir(),
    private readonly timerScale: TimerScale = DEFAULT_TIMER_SCALE,
  ) {
    this.psDir = psDir;
    if (Array.isArray(seed)) {
      this.seed = seed;
    } else if (seed !== undefined) {
      let value = BigInt.asUintN(64, BigInt(seed));
      const parts = Array.from({ length: 4 }, () => {
        value = BigInt.asUintN(64, value * 6364136223846793005n + 1442695040888963407n);
        return Number((value >> 16n) & 0xffffn);
      });
      this.seed = parts as [number, number, number, number];
    }
  }

  async run(
    agents: Record<Pid, BattleAgent>,
    onUpdate?: (lines: string[], publicLines: string[]) => void,
    signal?: AbortSignal,
    recovery?: RecoveryGate,
  ): Promise<BattleOutcome> {
    const { BattleStream } = loadShowdown(this.psDir);
    const stream = new BattleStream({ noCatch: true }) as BattleStream;
    const state: RouteState = {
      pov: { p1: [], p2: [] },
      log: [],
      publicLog: [],
      pendingSplit: [],
      winner: null,
      turns: 0,
    };
    const povCursor: Record<Pid, number> = { p1: 0, p2: 0 };
    const errors: Record<Pid, number> = { p1: 0, p2: 0 };
    const fallbacks: Record<Pid, number> = { p1: 0, p2: 0 };
    const lastRequest: Record<Pid, BattleRequest | undefined> = { p1: undefined, p2: undefined };
    const requestAt: Record<Pid, number> = { p1: 0, p2: 0 };
    const retryCount: Record<Pid, number> = { p1: 0, p2: 0 };
    const pendingError: Record<Pid, string | undefined> = { p1: undefined, p2: undefined };
    const suppressRequest: Record<Pid, boolean> = { p1: false, p2: false };
    const pending = new Map<Pid, PendingAction>();
    signal?.throwIfAborted();
    let abortPromise: Promise<never> | undefined;
    let onAbort: (() => void) | undefined;
    if (signal) {
      const aborted = Promise.withResolvers<never>();
      abortPromise = aborted.promise;
      onAbort = () => aborted.reject(signal.reason);
      signal.addEventListener('abort', onAbort, { once: true });
    }

    const route = (lines: string[]) => {
      const before = state.log.length;
      const publicBefore = state.publicLog.length;
      routeUpdateLines(lines, state);
      if (onUpdate && (state.log.length > before || state.publicLog.length > publicBefore)) {
        onUpdate(state.log.slice(before), state.publicLog.slice(publicBefore));
      }
    };

    const timerEvent = (pid: Pid, event: TimerEvent) => {
      const action = pending.get(pid);
      if (action) {
        pending.delete(pid);
        agents[pid].abandonDecision?.();
      }
      if (event === 'autodefault') fallbacks[pid] += 1;
      const line = `|timer|${event}`;
      state.pov[pid].push(line);
      state.log.push(line);
      const spectatorLine = `|-vgctimeout|${pid}|${event}`;
      onUpdate?.([spectatorLine], [spectatorLine]);
    };

    const start: Record<string, unknown> = { formatid: this.format };
    if (this.seed) start.seed = this.seed;
    stream.write(`>start ${JSON.stringify(start)}`);
    for (const pid of ['p1', 'p2'] as const) stream.write(`>player ${pid} ${JSON.stringify(this.players[pid])}`);
    const timer = new TimerAdapter(this.format, stream, timerEvent, this.psDir, this.timerScale);
    for (const pid of ['p1', 'p2'] as const) timer.setPlayer(pid, this.players[pid].name);
    const recoveryChanged = (paused: boolean) => {
      const lines = paused ? timer.pauseForRecovery() : timer.resumeAfterRecovery();
      if (lines.length) onUpdate?.(lines, lines);
    };
    const removeRecoveryListener = recovery?.onChange((pause) => recoveryChanged(Boolean(pause)));
    if (recovery?.paused) recoveryChanged(true);

    const schedule = (pid: Pid, request: BattleRequest, error?: string) => {
      if (pending.has(pid)) throw new Error(`received another request while ${pid} is still deciding`);
      const elapsed = error && request.timer ? (performance.now() - requestAt[pid]) / 1000 : 0;
      const currentRequest =
        request.timer && elapsed > 0
          ? {
              ...request,
              timer: {
                ...(request.timer.seconds === undefined
                  ? {}
                  : { seconds: Math.max(0, request.timer.seconds - elapsed) }),
                ...(request.timer.turnSeconds === undefined
                  ? {}
                  : { turnSeconds: Math.max(0, request.timer.turnSeconds - elapsed) }),
              },
            }
          : request;
      if (currentRequest.timer) {
        const line = `|-vgctimer|${pid}|${currentRequest.timer.seconds ?? ''}|${currentRequest.timer.turnSeconds ?? ''}`;
        onUpdate?.([line], [line]);
      } else {
        const line = `|-vgcdeciding|${pid}`;
        onUpdate?.([line], [line]);
      }
      const lines = state.pov[pid].slice(povCursor[pid]);
      povCursor[pid] = state.pov[pid].length;
      const token = Symbol(pid);
      let action: Promise<string>;
      try {
        action = Promise.resolve(agents[pid].act(currentRequest, { povLines: lines, ...(error ? { error } : {}) }));
      } catch (caught) {
        action = Promise.reject(caught);
      }
      const promise = action.then<ActionEvent, ActionEvent>(
        (choice) => ({ kind: 'action', pid, token, choice }),
        (caught) => ({ kind: 'action', pid, token, error: caught }),
      );
      pending.set(pid, { token, promise });
    };

    const handleSideUpdate = (lines: string[]) => {
      const pid = lines[0] as Pid;
      if (pid !== 'p1' && pid !== 'p2') return;
      for (const line of lines.slice(1)) {
        if (line.startsWith('|error|')) {
          errors[pid] += 1;
          retryCount[pid] += 1;
          pendingError[pid] = line;
          if (retryCount[pid] >= 3) {
            void Promise.resolve(timer.choose(pid, 'default')).catch(() => {});
            const stopLine = `|-vgctimerstop|${pid}`;
            onUpdate?.([stopLine], [stopLine]);
            fallbacks[pid] += 1;
            suppressRequest[pid] = line.includes('[Unavailable choice]');
            pendingError[pid] = undefined;
          } else if (!line.includes('[Unavailable choice]') && lastRequest[pid]) {
            schedule(pid, lastRequest[pid], line);
            pendingError[pid] = undefined;
          }
        } else if (line.startsWith('|request|')) {
          const payload = line.slice(9);
          if (!payload) continue;
          const request = JSON.parse(payload) as BattleRequest;
          lastRequest[pid] = request;
          requestAt[pid] = performance.now();
          if (suppressRequest[pid]) {
            suppressRequest[pid] = false;
            continue;
          }
          const error = pendingError[pid];
          if (!error) retryCount[pid] = 0;
          pendingError[pid] = undefined;
          if (!request.wait) schedule(pid, request, error);
        }
      }
    };

    let ended = false;
    let nextStream = streamEvent(stream);
    try {
      while (!ended) {
        const candidates: Array<Promise<StreamEvent | ActionEvent>> = [
          nextStream,
          ...[...pending.values()].map((item) => item.promise),
        ];
        if (abortPromise) candidates.push(abortPromise);
        let idleTimer: NodeJS.Timeout | undefined;
        if (!pending.size) {
          const { promise, resolve } = Promise.withResolvers<StreamEvent | ActionEvent>();
          idleTimer = setTimeout(() => resolve({ kind: 'stream', message: null }), 60_000);
          candidates.push(promise);
        }
        const event = await Promise.race(candidates);
        clearTimeout(idleTimer);
        if (event.kind === 'action') {
          const current = pending.get(event.pid);
          if (!current || current.token !== event.token) continue;
          pending.delete(event.pid);
          if (event.error !== undefined) throw event.error;
          await timer.choose(event.pid, event.choice ?? '');
          const stopLine = `|-vgctimerstop|${event.pid}`;
          onUpdate?.([stopLine], [stopLine]);
          continue;
        }
        if (event.message === null) throw new Error('simulator produced no output for 60 seconds');
        nextStream = streamEvent(stream);
        const message = timer.receive(event.message);
        const lines = message.split('\n');
        if (lines[0] === 'update') route(lines.slice(1));
        else if (lines[0] === 'sideupdate') handleSideUpdate(lines.slice(1));
        else if (lines[0] === 'end') {
          if (lines[1]) {
            const data = JSON.parse(lines.slice(1).join('\n')) as { winner?: string; turns?: number };
            state.winner = data.winner || state.winner;
            state.turns = data.turns ?? state.turns;
          }
          ended = true;
        } else if (lines[0]?.startsWith('|')) route(lines);
      }
    } finally {
      timer.end();
      removeRecoveryListener?.();
      if (signal && onAbort) signal.removeEventListener('abort', onAbort);
      await stream.writeEnd();
      for (const pid of pending.keys()) agents[pid].abandonDecision?.();
      pending.clear();
    }

    for (const pid of ['p1', 'p2'] as const) {
      const remaining = state.pov[pid].slice(povCursor[pid]);
      if (remaining.length) await agents[pid].observe(remaining);
    }
    return { winner: state.winner, turns: state.turns, log: state.log, pov: state.pov, errors, fallbacks };
  }
}
